fvm/transactionStorageLimiter.go

```diff
@@ -14,33 +14,73 @@ import (
 
 type TransactionStorageLimiter struct{}
 
-func NewTransactionStorageLimiter() TransactionStorageLimiter {
-	return TransactionStorageLimiter{}
-}
-
+// CheckStorageLimits checks each account that had its storage written to during a transaction, that its storage used
+// is less than its storage capacity.
+// Storage used is an FVM register and is easily accessible.
+// Storage capacity is calculated by the FlowStorageFees contract from the account's flow balance.
+//
+// The payers balance is considered to be maxTxFees lower that its actual balance, due to the fact that
+// the fee deduction step happens after the storage limit check.
 func (d TransactionStorageLimiter) CheckStorageLimits(
 	env environment.Environment,
 	addresses []flow.Address,
+	payer flow.Address,
+	maxTxFees uint64,
 ) error {
 	if !env.LimitAccountStorage() {
 		return nil
 	}
 
 	defer env.StartSpanFromRoot(trace.FVMTransactionStorageUsedCheck).End()
 
+	err := d.checkStorageLimits(env, addresses, payer, maxTxFees)
+	if err != nil {
+		return fmt.Errorf("storage limit check failed: %w", err)
+	}
+	return nil
+}
+
+func (d TransactionStorageLimiter) checkStorageLimits(
+	env environment.Environment,
+	addresses []flow.Address,
+	payer flow.Address,
+	maxTxFees uint64,
+) error {
 	commonAddresses := make([]common.Address, len(addresses))
```

Comment:
You probably want to initialize both of them to len+1 otherwise common addresses will have out of bounds access in case of `commonAddressesContainPayer`.